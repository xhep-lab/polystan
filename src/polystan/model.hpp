#ifndef POLYSTAN_MODEL_HPP_
#define POLYSTAN_MODEL_HPP_

#include <filesystem>
#include <limits>
#include <regex>
#include <string>
#include <utility>
#include <vector>

#include "polystan/read.hpp"
#include "polystan/json.hpp"
#include "polystan/read_err.hpp"
#include "polystan/version.hpp"
#include "polystan/metadata.hpp"
#include "polystan/mpi.hpp"

#include "bridgestan/src/bridgestan.h"
#include "polychord/interfaces.hpp"

namespace polystan {

std::optional<std::string> unconstrain_err(bs_model* model,
                                           std::vector<double> theta) {
  std::vector<double> theta_unc(theta.size());
  char* err;
  const int err_code
      = bs_param_unconstrain(model, theta.data(), theta_unc.data(), &err);
  return err_code == 0 ? std::nullopt
                       : std::optional<std::string>(add_to_err(err));
}

bs_model* make_bs_model(const std::string& data_file_name, unsigned int seed) {
  char* err;

  bs_model* model = bs_model_construct(data_file_name.c_str(), seed, &err);

  if ((model == nullptr) && (err != nullptr)) {
    std::string err_msg = add_to_err(err);
    if (data_file_name.empty()) {
      err_msg += "\nDid your model require a data file to be specified e.g. by data --file=your-data-file.json?\n";
    }
    throw std::runtime_error(err_msg);
  }

  return model;
}

bs_rng* make_bs_rng(bs_model* model, unsigned int seed) {
  if (bs_param_num(model, false, true) == bs_param_num(model, false, false)) {
    return nullptr;
  }

  char* err;
  bs_rng* rng = bs_rng_construct(seed, &err);

  if ((rng == nullptr) && (err != nullptr)) {
    throw std::runtime_error(add_to_err(err));
  }

  return rng;
}

double loglike(bs_model* model, bs_rng* rng, double* theta, int ndim,
               double* phi, int nderived) {
  int err_code = 0;
  char* err;

  // compute unconstrained parameters

  std::vector<double> theta_unc(ndim);
  err_code = bs_param_unconstrain(model, theta, theta_unc.data(), &err);

  if (err_code != 0) {
    throw std::runtime_error(add_to_err(err));
  }

  // constrain parameters to compute derived

  std::vector<double> theta_phi(ndim + nderived);
  err_code = bs_param_constrain(model, true, rng != nullptr, theta_unc.data(),
                                theta_phi.data(), rng, &err);

  if (err_code != 0) {
    throw std::runtime_error(add_to_err(err));
  }

  for (int i = 0; i < nderived; i++) {
    phi[i] = theta_phi[i + ndim];
  }

  // compute density - stan works on unconstrained space

  double loglike;
  err_code
      = bs_log_density(model, false, false, theta_unc.data(), &loglike, &err);

  if (err_code != 0) {
    throw std::runtime_error(add_to_err(err));
  }

  return loglike;
}

class Model {
 public:
  Model(const std::string& data_file_name, unsigned int seed, Settings settings)
      : seed(seed),
        data_file_name(data_file_name),
        model(make_bs_model(data_file_name, seed)),
        rng(make_bs_rng(model, seed)),
        settings(std::move(settings)) {
    check_unit_hypercube();
    fix_settings();
  }

  void check_unit_hypercube() const {
    const std::string msg
        = "\nParameters are not defined on unit hypercube; "
          "expect e.g. real<lower=0, upper=1>\n";

    const std::vector<double> zeros(ndims(), 0.);
    const auto zeros_err = unconstrain_err(model, zeros);
    if (zeros_err.has_value()) {
      throw std::runtime_error(zeros_err.value() + msg);
    }

    for (int i = 0; i < ndims(); i++) {
      std::vector<double> one(ndims(), 0.);
      one[i] = 1.;
      const auto one_err = unconstrain_err(model, one);
      if (one_err.has_value()) {
        throw std::runtime_error(one_err.value() + msg);
      }

      std::vector<double> gt_one(ndims(), 0.5);
      gt_one[i] = 1. + std::numeric_limits<double>::round_error();
      const auto gt_err = unconstrain_err(model, gt_one);
      if (!gt_err.has_value()) {
        throw std::runtime_error("> 1 was not out of bounds for parameter "
                                 + param_names()[i] + msg);
      }

      std::vector<double> lt_zero(ndims(), 0.5);
      lt_zero[i] = -std::numeric_limits<double>::round_error();
      const auto lt_err = unconstrain_err(model, lt_zero);
      if (!lt_err.has_value()) {
        throw std::runtime_error("< 0 was not out of bounds for parameter "
                                 + param_names()[i] + msg);
      }
    }
  }

  void run() const {
    std::filesystem::create_directory(settings.base_dir);

    static bs_model* model_(model);
    static bs_rng* rng_(rng);

    const auto this_loglike
        = [](double* theta, int ndim, double* phi, int nderived) {
            return loglike(model_, rng_, theta, ndim, phi, nderived);
          };

#ifdef USE_MPI
    run_polychord(this_loglike, settings, mpi::get_comm());
#else
    run_polychord(this_loglike, settings);
#endif
  }

  void write(const std::string& json_file_name) const {
    // polystan metadata

    json::Object polystan;

    polystan.add("polystan version", version);
    polystan.add("stan file name", stan_file_name);
    polystan.add("stan data file", data_file_name);
    polystan.add("stan build info", stan_build_info());
    polystan.add("seed", seed);

    // polychord metadata

    json::Object polychord;

    polychord.add("polychord version", polychord_version);
    polychord.add("nlive", settings.nlive);
    polychord.add("num_repeats", settings.num_repeats);
    polychord.add("precision_criterion", settings.precision_criterion);
    polychord.add("seed", settings.seed);

    // evidence

    json::Object evidence_;

    if (settings.write_stats) {
      const auto [logz, error_logz] = evidence();

      evidence_.add("metadata", "The evidence is log-normally distributed");
      evidence_.add("log evidence", logz);
      evidence_.add("error log evidence", error_logz);
    } else {
      evidence_.set("did not write stats file");
    }

    // samples

    json::Object samples_;

    if (settings.equals) {
      auto names_ = names();
      names_.insert(names_.begin(), "log likelihood");

      samples_.add("metadata", "These samples are equally weighted");
      json::Object data;
      data.add(names_, samples());
      samples_.add("data", data);
    } else {
      samples_.set("did not write equally weighted points");
    }

    // write to disk

    json::Object document;
    document.add("polystan", polystan);
    document.add("polychord", polychord);
    document.add("evidence", evidence_);
    document.add("samples", samples_);
    document.write(json_file_name);
  }

  const std::string data_file_name;
  const std::string stan_file_name{polystan::stan_file_name};

  std::string stan_build_info() const {
    std::string build_info = bs_model_info(model);
    build_info = std::regex_replace(build_info, std::regex("\n"), ", ");
    build_info = std::regex_replace(build_info, std::regex("\t"), "");
    return build_info;
  }

  std::vector<std::string> names() const {
    return read::param_names(bs_param_names(model, true, true));
  }

  std::vector<std::string> param_names() const {
    std::vector<std::string> names_ = names();
    return std::vector<std::string>(names_.begin(), names_.begin() + ndims());
  }

  std::vector<std::string> derived_names() const {
    std::vector<std::string> names_ = names();
    return std::vector<std::string>(names_.begin() + ndims(), names_.end());
  }

  int ndims() const { return bs_param_num(model, false, false); }

  int nderived() const {
    return bs_param_num(model, true, true) - bs_param_num(model, false, false);
  }

  std::string basename() const {
    return std::filesystem::weakly_canonical(settings.base_dir)
           / settings.file_root;
  }

  std::vector<std::vector<double>> samples() const {
    return read::samples(basename() + "_equal_weights.txt");
  }

  std::array<double, 2> evidence() const {
    return read::evidence(basename() + ".stats");
  }

 private:
  void fix_settings() {
    settings.nDims = ndims();
    settings.nDerived = nderived();

    const Settings default_zero(0, 0);
    const Settings fixed(settings.nDims, settings.nDerived);

    if (settings.grade_dims == default_zero.grade_dims) {
      settings.grade_dims = fixed.grade_dims;
    }

    if (settings.num_repeats == default_zero.num_repeats) {
      settings.num_repeats = fixed.num_repeats;
    }
  }

  bs_model* model;
  bs_rng* rng;
  Settings settings;
  unsigned int seed;
};

}  // end namespace polystan

#endif  // POLYSTAN_MODEL_HPP_
