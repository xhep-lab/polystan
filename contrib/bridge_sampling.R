# run bridge sampling with Stan MCMC chains on Stan models

dir.create(Sys.getenv("R_LIBS_USER"), recursive = TRUE)
.libPaths(Sys.getenv("R_LIBS_USER"))

if (!require("pacman")) install.packages("pacman")
pacman::p_load("bridgesampling", "rstan", "jsonlite")


bridge_sampling = function(target, iter, warmup, chains, seed, headers) {
  set.seed(seed)
  rstan_options(auto_write = TRUE)
  wd <- getwd()
  setwd(headers)

  stanfit <- stan(file = paste(target, ".stan", sep = ""),
                  data = fromJSON(paste(target, ".data.json", sep = "")),
                  iter = iter,
                  warmup = warmup,
                  chains = chains,
                  seed = seed)
        
  setwd(wd)
        
  sampler_params <- get_sampler_params(stanfit, inc_warmup = TRUE)
  mcmc_neval <- sum(do.call(rbind, sampler_params)[,'n_leapfrog__'])

  result <- bridge_sampler(stanfit)
  bs_neval <- length(result$q11) + length(result$q21)
  neval <- mcmc_neval + bs_neval
  logz <- result$logml
  err_logz <- sqrt(error_measures(result)$re2)
  return(c(logz, err_logz, neval))
}
