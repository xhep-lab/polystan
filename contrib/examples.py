"""
Find example programs
=====================
"""

import os


CWD = os.path.dirname(os.path.realpath(__file__))
ROOT = os.path.normpath(os.path.join(CWD, ".."))
EXAMPLE_DIR = os.path.join(ROOT, "examples")


def search_programs(dir_):
    return [
        os.path.join(dir_, n)
        for n in os.listdir(dir_)
        if n.endswith(".stan")
    ]


def examples():
    return search_programs(EXAMPLE_DIR)


def example(name):
    stan_file = os.path.join(EXAMPLE_DIR, name)
    assert os.path.isfile(stan_file)
    return stan_file
