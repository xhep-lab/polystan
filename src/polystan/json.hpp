#ifndef POLYSTAN_JSON_HPP_
#define POLYSTAN_JSON_HPP_

#include <rapidjson/document.h>
#include <rapidjson/ostreamwrapper.h>
#include <rapidjson/prettywriter.h>

#include <string>
#include <fstream>
#include <sstream>
#include <utility>
#include <vector>

namespace polystan {
namespace json {

namespace rj = rapidjson;
typedef rapidjson::MemoryPoolAllocator<> Alloc;

Alloc& get_allocator() {
  static rj::GenericDocument<rj::UTF8<>> doc;
  return doc.GetAllocator();
}

template <typename T>
rj::Value Vector(const std::vector<T>& vec, Alloc& alloc) {
  rj::Value value(rj::kArrayType);
  value.Reserve(vec.size(), alloc);
  for (const T& elem : vec) {
    value.PushBack(elem, alloc);
  }
  return value;
}

rj::Value String(const std::string& str, Alloc& alloc) {
  rj::Value value(rj::kStringType);
  value.SetString(str.c_str(), str.size(), alloc);
  return value;
}

class Object {
 public:
  Object() : value(rj::kObjectType), alloc(get_allocator()) {}

  template <typename T>
  void add(const std::string& name, const std::vector<T>& data) {
    value.AddMember(String(name, alloc).Move(), Vector(data, alloc).Move(),
                    alloc);
  }

  template <typename T>
  void add(const std::string& name, T data) {
    value.AddMember(String(name, alloc).Move(), data, alloc);
  }

  void add(const std::string& name, const char* data) {
    value.AddMember(String(name, alloc).Move(), String(data, alloc).Move(),
                    alloc);
  }

  void add(const std::string& name, const std::string& data) {
    add(name, data.c_str());
  }

  template <typename T>
  void add(std::vector<std::string> names, std::vector<T> data) {
    for (int i = 0; i < names.size(); i++) {
      add(names[i], data[i]);
    }
  }

  void add(const std::string& name, Object& child) {
    value.AddMember(String(name, alloc).Move(), child.value.Move(), alloc);
  }

  template <typename T>
  void set(T value_) {
    value = value_;
  }

  void set(const std::string& value_) {
    value.SetString(value_.c_str(), value_.size(), alloc);
  }

  void set(const char* value_) { set(std::string(value_)); }

  void write(const std::string& json_file_name) {
    std::ofstream ofs(json_file_name);
    rj::OStreamWrapper osw(ofs);
    rj::PrettyWriter<rj::OStreamWrapper> writer(osw);
    value.Accept(writer);
  }

 private:
  Alloc& alloc;
  rj::Value value;
};

}  // end namespace json
}  // end namespace polystan

#endif  // POLYSTAN_JSON_HPP_
