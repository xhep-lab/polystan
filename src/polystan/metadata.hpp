#ifndef POLYSTAN_METADATA_HPP_
#define POLYSTAN_METADATA_HPP_

namespace polystan {
extern const char* stan_file_name;
extern const char* stan_model_name;
}  // end namespace polystan

#endif  // POLYSTAN_METADATA_HPP_
