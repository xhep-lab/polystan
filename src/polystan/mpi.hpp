#ifndef POLYSTAN_MPI_HPP_
#define POLYSTAN_MPI_HPP_

#ifdef USE_MPI
#include <mpi.h>
#endif

namespace polystan {
namespace mpi {

void initialize() {
#ifdef USE_MPI
  MPI_Init(NULL, NULL);
#endif
}

void finalize() {
#ifdef USE_MPI
  MPI_Finalize();
#endif
}

#ifdef USE_MPI
MPI_Comm dup_comm() {
  MPI_Comm comm;
  MPI_Comm_dup(MPI_COMM_WORLD, &comm);
  return comm;
}

MPI_Comm& get_comm() {
  static MPI_Comm comm = dup_comm();
  return comm;
}

int get_size() {
  int size;
  MPI_Comm_size(get_comm(), &size);
  return size;
}
#endif

void barrier() {
#ifdef USE_MPI
  MPI_Barrier(get_comm());
#endif
}

bool is_rank_zero() {
#ifdef USE_MPI
  int rank;
  MPI_Comm_rank(get_comm(), &rank);
  return rank == 0;
#else
  return true;
#endif
}

}  // end namespace mpi
}  // end namespace polystan

#endif  // POLYSTAN_MPI_HPP_
