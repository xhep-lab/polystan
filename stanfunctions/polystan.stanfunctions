// inverse transform sampling from $x \sim \mathcal{U}(0, 1)$

vector flat_prior(vector x, real a, real b) {
  // if $x \sim \mathcal{U}(0, 1)$ then $y \sim \mathcal{U}(a, b)$
  return (b - a) * x + a;
}

vector log_prior(vector x, real a, real b) {
  // if $x \sim \mathcal{U}(0, 1)$ then $\log y \sim \mathcal{U}(\log a, \log b)$
  return a * exp(log(b / a) * x);
}

vector std_normal_prior(vector x) {
  // if $x \sim \mathcal{U}(0, 1)$ then $y \sim \mathcal{N}(0, 1)$
  return inv_Phi(x);
}

vector half_std_normal_prior(vector x) {
  // if $x \sim \mathcal{U}(0, 1)$ then $y \sim \mathcal{N}_{>0}(0, 1)$
  return std_normal_prior(flat_prior(x, 0.5, 1.));
}

vector normal_prior(vector x, real mu, real sigma) {
  // if $x \sim \mathcal{U}(0, 1)$ then $y \sim \mathcal{N}(\mu, \sigma^2)$
  return inv_Phi(x) * sigma + mu;
}

vector half_normal_prior(vector x, real mu, real sigma) {
  // if $x \sim \mathcal{U}(0, 1)$ then $y \sim \mathcal{N}_{>\mu}(\mu, \sigma^2)$
  return normal_prior(flat_prior(x, 0.5, 1.), mu, sigma);
}

vector multi_normal_prior(vector x, vector mu, matrix Sigma) {
  // if $\vec x \sim \mathcal{U}(0, 1)$ then $\vec y \sim \mathcal{N}(\vec \mu, \vec \Sigma)$
  return multi_normal_cholesky_prior(x, mu, cholesky_decompose(Sigma));
}

vector multi_normal_cholesky_prior(vector x, vector mu, matrix L) {
  // if $\vec x \sim \mathcal{U}(0, 1)$ then $\vec y \sim \mathcal{N}(\vec \mu, \vec \Sigma = \vec L \vec L^T)$
  return L * inv_Phi(x) + mu;
}

vector log10normal_prior(vector x, real mu, real sigma) {
  // if $x \sim \mathcal{U}(0, 1)$ then $\log_{10} y \sim \mathcal{N}(\mu, \sigma^2)$
  return exp(log(10.) * normal_prior(x, mu, sigma));
}

vector lognormal_prior(vector x, real mu, real sigma) {
  // if $x \sim \mathcal{U}(0, 1)$ then $\log y \sim \mathcal{N}(\mu, \sigma^2)$
  return exp(normal_prior(x, mu, sigma));
}

real beta_prior(real x, real alpha, real beta) {
  // if $x \sim \mathcal{U}(0, 1)$ then $y \sim \text{Beta}(\alpha, \beta)$
  return inv_inc_beta(x, alpha, beta);
}

vector beta_prior(vector x, real alpha, real beta) {
  // if $x \sim \mathcal{U}(0, 1)$ then $y \sim \text{Beta}(\alpha, \beta)$
  vector[size(x)] y;
  for (i in 1 : size(x)) {
    y[i] = beta_prior(x[i], alpha, beta);
  }
  return y;
}

vector exponential_prior(vector x, real lambda) {
  // if $x \sim \mathcal{U}(0, 1)$ then $y \sim \text{Exp}(\lambda)$
  return -log(1. - x) / lambda;
}

vector cauchy_prior(vector x, real x0, real gamma) {
  // if $x \sim \mathcal{U}(0, 1)$ then $y \sim \text{Cauchy}(x_0, \gamma)$
  return x0 + gamma * tan(pi() * (x - 0.5));
}

vector half_cauchy_prior(vector x, real x0, real gamma) {
  // if $x \sim \mathcal{U}(0, 1)$ then $y \sim \text{Cauchy}_{>x_0}(x_0, \gamma)$
  return cauchy_prior(flat_prior(x, 0.5, 1.), x0, gamma);
}