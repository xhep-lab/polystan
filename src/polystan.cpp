#include <iostream>
#include <filesystem>
#include <string>

#include "polystan/splash.hpp"
#include "polystan/model.hpp"
#include "polystan/polychord_cli.hpp"
#include "polystan/version.hpp"
#include "polystan/mpi.hpp"

#include "CLI11/CLI11.hpp"
#include "polychord/interfaces.hpp"

namespace ps = polystan;


template<typename T>
class Display {
 public:
  Display(T data) : data(data) {}
  void operator() (int) {
    throw CLI::CallForVersion(data, CLI::ExitCodes::Success);
  }
 private:
  T data;  
};


int main(int argc, char** argv) {
    // parse cli

    CLI::App app("PolyStan built with " + std::string(ps::stan_file_name));

    // construct settings before knowing nDims & nDerived. they aren't knowable until
    // constructing the model as could depend on data file

    Settings settings(0, 0);

    // adjust some default settings

    settings.maximise = false;
    settings.write_live = true;
    settings.write_dead = true;
    settings.write_resume = true;
    settings.equals = true;
    settings.write_stats = true;
    settings.read_resume = true;
    settings.file_root = ps::stan_model_name;

    CLI::App* pc = app.add_subcommand("polychord", "PolyChord settings");
    ps::AddPolyChord(pc, &settings);

    CLI::App* data = app.add_subcommand("data", "Data settings");
    std::string data_file_name;
    data->add_option("--file", data_file_name, "Data file name")->check(CLI::ExistingFile);

    CLI::App* random = app.add_subcommand("random", "Control Stan random number generation");
    unsigned int seed = 0;
    random->add_option("--seed", seed, "Random seed")->check(CLI::PositiveNumber);

    CLI::App* output = app.add_subcommand("output", "Control Stan output");
    std::string json_file_name = std::string(ps::stan_model_name) + ".json";
    output->add_option("--file", json_file_name , "JSON file output name");

    app.set_version_flag("--version", ps::version);
    app.add_flag("--polychord-version", Display(ps::polychord_version), "Display PolyChord version information and exit");
    app.add_flag("--stan-file-name", Display(ps::stan_file_name), "Display Stan file name and exit");

    CLI11_PARSE(app, argc, argv);

    data_file_name = std::filesystem::weakly_canonical(data_file_name);
    json_file_name = std::filesystem::weakly_canonical(json_file_name);

    // invoke main program

    ps::mpi::initialize();

    const ps::Model model(data_file_name, seed, settings);

    if (ps::mpi::is_rank_zero()) {
      std::cout << ps::splash::start(model) << std::endl;
    }

    ps::mpi::barrier();
  
    model.run();

    // write results to disk in json format

    if (ps::mpi::is_rank_zero()) {
      model.write(json_file_name);
      std::cout << ps::splash::end(json_file_name, model) << std::endl;
    }

    ps::mpi::finalize();
}
