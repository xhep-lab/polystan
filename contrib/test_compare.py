"""
Compare bridge sampling and polystan
====================================
"""

import os
import sigfig
import pytest

from examples import examples
from polystan import run_polystan
from bridge_sampling import run_bridge_sampling


POLYCHORD_SETTINGS = {"no-feedback": True,
                      "no-derived": True, "nlive": 1000, "write-stats": True}
BRIDGE_SAMPLING_SETTINGS = {"iter_": 20_000, "warmup": 5_000, "chains": 4}

EXAMPLES = {os.path.basename(e): e for e in examples()}


@pytest.mark.parametrize("example", EXAMPLES.keys())
def test_bridge_sampling(example, snapshot):
    example = EXAMPLES[example]
    bridge_sampling = run_bridge_sampling(example, **BRIDGE_SAMPLING_SETTINGS)
    rounded = {k: sigfig.round(v, sigfigs=6) for k, v in bridge_sampling.items()}
    assert rounded == snapshot


@pytest.mark.parametrize("example", EXAMPLES.keys())
def test_polystan(example, snapshot):
    example = EXAMPLES[example]
    polystan = run_polystan(example, polychord=POLYCHORD_SETTINGS)
    polystan_summary = {"log evidence": polystan['sample_stats']['evidence'].data[0][0]['log evidence'],
                        "error log evidence": polystan['sample_stats']['evidence'].data[0][0]['error log evidence'],
                        "neval": polystan['sample_stats']['neval'].data[0][0]['neval']}
    rounded = {k: sigfig.round(v, sigfigs=6) for k, v in polystan_summary.items()}
    assert rounded == snapshot
