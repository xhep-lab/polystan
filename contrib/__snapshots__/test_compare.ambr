# serializer version: 1
# name: test_bridge_sampling[bernoulli.stan]
  dict({
    'error log evidence': 0.000361986,
    'log evidence': -6.20489,
    'neval': 243238,
  })
# ---
# name: test_bridge_sampling[disaster.stan]
  dict({
    'error log evidence': 0.00193918,
    'log evidence': -62.0608,
    'neval': 81476800,
  })
# ---
# name: test_bridge_sampling[eggbox.stan]
  dict({
    'error log evidence': 0.125202,
    'log evidence': -28.3233,
    'neval': 796773,
  })
# ---
# name: test_bridge_sampling[gaussian.stan]
  dict({
    'error log evidence': 0.000555094,
    'log evidence': 0.000191301,
    'neval': 563334,
  })
# ---
# name: test_bridge_sampling[glmm_h0.stan]
  dict({
    'error log evidence': 0.000499566,
    'log evidence': -156.479,
    'neval': 1104770,
  })
# ---
# name: test_bridge_sampling[glmm_h1.stan]
  dict({
    'error log evidence': 0.00474544,
    'log evidence': -156.713,
    'neval': 3374200,
  })
# ---
# name: test_bridge_sampling[himmelblau.stan]
  dict({
    'error log evidence': 0.0334683,
    'log evidence': -4.86578,
    'neval': 332999,
  })
# ---
# name: test_bridge_sampling[priors.stan]
  dict({
    'error log evidence': 0.000668779,
    'log evidence': -0.000498468,
    'neval': 405564,
  })
# ---
# name: test_bridge_sampling[rastrigin.stan]
  dict({
    'error log evidence': 0.0464627,
    'log evidence': -104.182,
    'neval': 568046,
  })
# ---
# name: test_bridge_sampling[rosenbrock.stan]
  dict({
    'error log evidence': 0.0244839,
    'log evidence': -9.37889,
    'neval': 2240930,
  })
# ---
# name: test_bridge_sampling[shell.stan]
  dict({
    'error log evidence': 0.012201,
    'log evidence': -5.70835,
    'neval': 3949480,
  })
# ---
# name: test_bridge_sampling[slab_spike.stan]
  dict({
    'error log evidence': 0.00071648,
    'log evidence': -5.34118,
    'neval': 259280,
  })
# ---
# name: test_compare[bernoulli.stan]
  dict({
    'bridge_sampling': dict({
      'error log evidence': 0.00036198557800761855,
      'log evidence': -6.204891875185858,
      'neval': 243238,
    }),
    'polystan': dict({
      'error log evidence': 0.028646834194660187,
      'log evidence': -6.20444393157959,
      'neval': 171591,
    }),
  })
# ---
# name: test_compare[disaster.stan]
  dict({
    'bridge_sampling': dict({
      'error log evidence': 0.0019391763933108972,
      'log evidence': -62.060800597355986,
      'neval': 81476776,
    }),
    'polystan': dict({
      'error log evidence': 0.0852375477552414,
      'log evidence': -176.81947326660156,
      'neval': 1019612,
    }),
  })
# ---
# name: test_compare[eggbox.stan]
  dict({
    'bridge_sampling': dict({
      'error log evidence': 0.12520157052391095,
      'log evidence': -28.32333641032787,
      'neval': 796773,
    }),
    'polystan': dict({
      'error log evidence': 0.10254349559545517,
      'log evidence': -15.058009147644043,
      'neval': 3616587,
    }),
  })
# ---
# name: test_compare[gaussian.stan]
  dict({
    'bridge_sampling': dict({
      'error log evidence': 0.0005550944925997108,
      'log evidence': 0.0001913011388094965,
      'neval': 563334,
    }),
    'polystan': dict({
      'error log evidence': 0.09363611787557602,
      'log evidence': 0.15033656358718872,
      'neval': 4504083,
    }),
  })
# ---
# name: test_compare[glmm_h0.stan]
  dict({
    'bridge_sampling': dict({
      'error log evidence': 0.000499565932972676,
      'log evidence': -156.4786688835977,
      'neval': 1104770,
    }),
    'polystan': dict({
      'error log evidence': 0.08078835159540176,
      'log evidence': -156.21998596191406,
      'neval': 604896,
    }),
  })
# ---
# name: test_compare[glmm_h1.stan]
  dict({
    'bridge_sampling': dict({
      'error log evidence': 0.004745438237276931,
      'log evidence': -156.7134162084404,
      'neval': 3374200,
    }),
    'polystan': dict({
      'error log evidence': 0.11430605500936508,
      'log evidence': -156.36294555664062,
      'neval': 10294516,
    }),
  })
# ---
# name: test_compare[himmelblau.stan]
  dict({
    'bridge_sampling': dict({
      'error log evidence': 0.03346829929570338,
      'log evidence': -4.8657796489451695,
      'neval': 332999,
    }),
    'polystan': dict({
      'error log evidence': 0.06811784952878952,
      'log evidence': -4.635522842407227,
      'neval': 912030,
    }),
  })
# ---
# name: test_compare[priors.stan]
  dict({
    'bridge_sampling': dict({
      'error log evidence': 0.0006687792506750713,
      'log evidence': -0.0004984683559818848,
      'neval': 405564,
    }),
    'polystan': dict({
      'error log evidence': 0.0009990012040361762,
      'log evidence': -0.0009999993490055203,
      'neval': 20579,
    }),
  })
# ---
# name: test_compare[rastrigin.stan]
  dict({
    'bridge_sampling': dict({
      'error log evidence': 0.04646268945472804,
      'log evidence': -104.18237881301465,
      'neval': 568046,
    }),
    'polystan': dict({
      'error log evidence': 0.1631445735692978,
      'log evidence': -24.148386001586914,
      'neval': 12564554,
    }),
  })
# ---
# name: test_compare[rosenbrock.stan]
  dict({
    'bridge_sampling': dict({
      'error log evidence': 0.02448387058417394,
      'log evidence': -9.378885257962041,
      'neval': 2240925,
    }),
    'polystan': dict({
      'error log evidence': 0.11285044997930527,
      'log evidence': -9.257025718688965,
      'neval': 2077030,
    }),
  })
# ---
# name: test_compare[shell.stan]
  dict({
    'bridge_sampling': dict({
      'error log evidence': 0.012200964559424895,
      'log evidence': -5.708345130782098,
      'neval': 3949480,
    }),
    'polystan': dict({
      'error log evidence': 0.09432794898748398,
      'log evidence': -5.640108108520508,
      'neval': 1847206,
    }),
  })
# ---
# name: test_compare[slab_spike.stan]
  dict({
    'bridge_sampling': dict({
      'error log evidence': 0.0007164800110462592,
      'log evidence': -5.341181197687126,
      'neval': 259280,
    }),
    'polystan': dict({
      'error log evidence': 0.050396740436553955,
      'log evidence': -4.599237442016602,
      'neval': 341354,
    }),
  })
# ---
# name: test_polystan[bernoulli.stan]
  dict({
    'error log evidence': 0.0286468,
    'log evidence': -6.20444,
    'neval': 171591,
  })
# ---
# name: test_polystan[disaster.stan]
  dict({
    'error log evidence': 0.0852375,
    'log evidence': -176.819,
    'neval': 1019610,
  })
# ---
# name: test_polystan[eggbox.stan]
  dict({
    'error log evidence': 0.102543,
    'log evidence': -15.058,
    'neval': 3616590,
  })
# ---
# name: test_polystan[gaussian.stan]
  dict({
    'error log evidence': 0.0936361,
    'log evidence': 0.150337,
    'neval': 4504080,
  })
# ---
# name: test_polystan[glmm_h0.stan]
  dict({
    'error log evidence': 0.0807884,
    'log evidence': -156.22,
    'neval': 604896,
  })
# ---
# name: test_polystan[glmm_h1.stan]
  dict({
    'error log evidence': 0.114306,
    'log evidence': -156.363,
    'neval': 10294500,
  })
# ---
# name: test_polystan[himmelblau.stan]
  dict({
    'error log evidence': 0.0681178,
    'log evidence': -4.63552,
    'neval': 912030,
  })
# ---
# name: test_polystan[priors.stan]
  dict({
    'error log evidence': 0.000999001,
    'log evidence': -0.000999999,
    'neval': 20579,
  })
# ---
# name: test_polystan[rastrigin.stan]
  dict({
    'error log evidence': 0.163145,
    'log evidence': -24.1484,
    'neval': 12564600,
  })
# ---
# name: test_polystan[rosenbrock.stan]
  dict({
    'error log evidence': 0.11285,
    'log evidence': -9.25703,
    'neval': 2077030,
  })
# ---
# name: test_polystan[shell.stan]
  dict({
    'error log evidence': 0.0943279,
    'log evidence': -5.64011,
    'neval': 1847210,
  })
# ---
# name: test_polystan[slab_spike.stan]
  dict({
    'error log evidence': 0.0503967,
    'log evidence': -4.59924,
    'neval': 341354,
  })
# ---
